src/BSON/ArrayList.c

```diff
@@ -0,0 +1,287 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <php.h>
+#include <ext/standard/base64.h>
+#include <Zend/zend_interfaces.h>
+#include <Zend/zend_operators.h>
+#include <ext/standard/php_var.h>
+#include <zend_smart_str.h>
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "php_phongo.h"
+#include "phongo_error.h"
+#include "phongo_bson_encode.h"
+#include "BSON/ArrayList_arginfo.h"
+#include "BSON/Iterator.h"
+
+zend_class_entry* php_phongo_arraylist_ce;
+
+/* Initialize the object and return whether it was successful. An exception will
+ * be thrown on error. */
+static bool php_phongo_arraylist_init(php_phongo_arraylist_t* intern, bson_t* bson)
+{
+	intern->bson = bson;
+
+	return true;
+}
+
+static HashTable* php_phongo_arraylist_get_properties_hash(phongo_compat_object_handler_type* object, bool is_temp)
+{
+	php_phongo_arraylist_t* intern;
+	HashTable*              props;
+
+	intern = Z_OBJ_ARRAYLIST(PHONGO_COMPAT_GET_OBJ(object));
+
+	PHONGO_GET_PROPERTY_HASH_INIT_PROPS(is_temp, intern, props, 2);
+
+	if (!intern->bson) {
+		return props;
+	}
+
+	{
+		zval data, length;
+
+		ZVAL_STRINGL(&data, (const char*) bson_get_data(intern->bson), intern->bson->len);
+		zend_hash_str_update(props, "data", sizeof("data") - 1, &data);
+
+		ZVAL_LONG(&length, intern->bson->len);
+		zend_hash_str_update(props, "length", sizeof("length") - 1, &length);
+	}
+
+	return props;
+}
+
+PHONGO_DISABLED_CONSTRUCTOR(MongoDB_BSON_ArrayList)
+
+static PHP_METHOD(MongoDB_BSON_ArrayList, fromPHP)
+{
+	zval                    zv;
+	php_phongo_arraylist_t* intern;
+	zval*                   data;
+
+	PHONGO_PARSE_PARAMETERS_START(1, 1)
+	PHONGO_PARAM_ARRAY_OR_OBJECT(data)
+	PHONGO_PARSE_PARAMETERS_END();
+
+	if (!zend_array_is_list(Z_ARRVAL_P(data))) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT, "Expected value to be a list, but given array is not.");
+		return;
+	}
+
+	object_init_ex(&zv, php_phongo_arraylist_ce);
+	intern = Z_ARRAYLIST_OBJ_P(&zv);
+
+	intern->bson = bson_new();
+	php_phongo_zval_to_bson(data, PHONGO_BSON_NONE, intern->bson, NULL);
+
+	RETURN_ZVAL(&zv, 1, 1);
+}
+
+static bool seek_iter_to_index(bson_iter_t* iter, zend_long index)
+{
+	for (zend_long i = 0; i <= index; i++) {
+		if (!bson_iter_next(iter)) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static PHP_METHOD(MongoDB_BSON_ArrayList, get)
+{
+	php_phongo_arraylist_t* intern;
+	zend_long               index;
+	bson_iter_t             iter;
+
+	PHONGO_PARSE_PARAMETERS_START(1, 1)
+	Z_PARAM_LONG(index)
+	PHONGO_PARSE_PARAMETERS_END();
+
+	intern = Z_ARRAYLIST_OBJ_P(getThis());
+	if (!bson_iter_init(&iter, intern->bson)) {
+		phongo_throw_exception(PHONGO_ERROR_RUNTIME, "Could not initialize BSON iterator.");
+	}
+
+	if (!seek_iter_to_index(&iter, index)) {
+		RETURN_NULL();
```

Comment:
Would it make more sense to throw an exception here? Otherwise, it's ambiguous if the index exists and contains BSON `null` or does not exist. I'm not going to suggest we create an equivalent of [OutOfBoundsException](https://www.php.net/manual/en/class.outofboundsexception.php); a RuntimeException would suffice.

Reply:
For an ArrayList, we could throw an exception as we're not expecting gaps or anything, but rather an ordered list that users traverse. I assume that users won't be accessing indexes randomly here (as there's no way of knowing the number of elements without counting them before). I'm also fine with throwing a regular `OutOfBoundsException` without creating our own specific class. That would be more specific than using a `RuntimeException`.

Reply:
All exceptions thrown by the extension implement our Exception interface, so it'd be inconsistent to start throwing an SPL class directly. I'm also can't imagine that many users care about the granularity of our exception classes, so I wouldn't push for introducing our own OutOfBoundsException just for this case.

Happy to defer to you, though.

Reply:
This now throws on a non-existent key.